astropy/utils/decorators.py

```diff
@@ -1,14 +1,20 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """Sundry function and class decorators."""
 
+from __future__ import annotations
+
 import functools
 import inspect
 import textwrap
 import threading
 import types
 import warnings
+from collections.abc import Callable
+from functools import partial, update_wrapper
 from inspect import signature
+from typing import Any
 
+from ._signature_deprecations import _depr_kws_wrap
```

Comment:
I am not familiar with cosmology code base, so please explain to me why deprecating `z` in cosmology needs C extension in `utils`. Thanks!

Reply:
Ah, I just saw https://github.com/astropy/astropy/pull/15859#issuecomment-2180747472 . Performance boost is always welcome but I am afraid this C code would also make the decorator hard to maintain going forward. I will put this on queue to review.

Reply:
Oh, so this boost is only for a very specific case. And I feel like the new function (ignoring performance) duplicated something `deprecated_renamed_argument` can already do. I need to ponder.

Even if we decide to keep this as-is, license for https://pypi.org/project/positional-defaults/ would need to be included in https://github.com/astropy/astropy/tree/main/licenses

Reply:
Also the new function is technically new API, so might also need a change log for `utils`.

Reply:
> And I feel like the new function (ignoring performance) duplicated something `deprecated_renamed_argument` can already do.

I don't think that's the case. `deprecated_renamed_argument` only works if a keyword argument is still allowed, albeit under a new name, while `deprecated_keyword` forces the user to stop using the argument as keyword.

> I am afraid this C code would also make the decorator hard to maintain going forward
> Also the new function is technically new API

Good points. We don't actually need to make it public API. Maybe we can keep it private for now and see where it goes ? If it's not used at all after this deprecation cycle is concluded we can just get rid of this C extension altogether ?

> Even if we decide to keep this as-is, license for [pypi.org/project/positional-defaults](https://pypi.org/project/positional-defaults/) would need to be included in [main/licenses](https://github.com/astropy/astropy/tree/main/licenses?rgh-link-date=2024-06-24T19%3A28%3A54Z)

Thanks !

Reply:
> Good points. We don't actually need to make it public API

Agreed. Let's make it private. If @mhvk or @pllim prefers, we can also move it into `cosmology`. Though looking around `Astropy` I would recommend other maintainers avail themselves of this useful utility and improve API by making some required arguments positional-only.